### PR TITLE
feat: Extend unfinished rental if lessor and tenant are the same

### PR DESCRIPTION
**Pull request by fzavalia**

### CODE REVIEW

File: contracts/Rentals.sol [hunk 1]
```diff
@@ -396,19 +393,37 @@ contract Rentals is NonceVerifiable, NativeMetaTransaction, IERC721Receiver {
...
+        // True if the asset is currently rented.
+        bool rented = isRented(_contractAddress, _tokenId);
+        // True if the asset rental period is over, but is has not been claimed back from the contract.
+        bool reRent = !rented && rental.lessor != address(0);
+        // True if the asset rental period is not over yet, but the lessor and the tenant are the same.
+        bool extend = rented && rental.lessor == _lessor && rental.tenant == _tenant;
+
+        if (!extend && !reRent) {
+            // Verify that the asset is not already rented.
+            require(!rented, "Rentals#_rent: CURRENTLY_RENTED");
```
Comment (nachomazzara) [inline contracts/Rentals.sol:405]: What do you think of moving the if into something like

```
require(extend || reRent ||  !rented, "Rentals#_rent: CURRENTLY_RENTED");
```

consume less gas?

File: contracts/Rentals.sol [hunk 2]
```diff
@@ -420,7 +435,7 @@ contract Rentals is NonceVerifiable, NativeMetaTransaction, IERC721Receiver {
         }
 
         // Only transfer the ERC721 to this contract if it doesn't already have it.
-        if (_ownerOf(address(asset), _tokenId) != address(this)) {
+        if (!extend && !reRent && _ownerOf(address(asset), _tokenId) != address(this)) {
```
Comment (nachomazzara) [inline contracts/Rentals.sol:438]: Why do we need to check for these variables? Is not ok only to check the ownership with `_ownerOf` as before?

File: contracts/Rentals.sol [hunk 3]
```diff
@@ -396,19 +393,37 @@ contract Rentals is NonceVerifiable, NativeMetaTransaction, IERC721Receiver {
...
-        }
 
-        // Set the rental finish timestamp in the rentals mapping.
-        rental.endDate = block.timestamp + _rentalDays * 86400; // 86400 = seconds in a day
+            // Track the new tenant in the mapping.
+            rental.tenant = _tenant;
 
-        // Track the new tenant in the mapping.
-        rental.tenant = _tenant;
+            // Set te end date of the rental according to the provided rental days
```
Comment (nachomazzara) [inline contracts/Rentals.sol:424]:
```suggestion
            // Set the end date of the rental according to the provided rental days
```

File: test/Rentals.spec.ts [hunk 4]
```diff
@@ -1732,6 +1801,73 @@ describe('Rentals', () => {
...
+          acceptListingParams.rentalDays,
+          acceptListingParams.fingerprint
+        )
+
+      rental = await rentals.rentals(listingParams.contractAddress, listingParams.tokenId)
+
+      endDate += daysToSeconds(acceptListingParams.rentalDays)
+
+      expect(rental.endDate).to.equal(endDate)
+    })
```
Comment (nachomazzara) [inline test/Rentals.spec.ts:1869]: I would add more tests for revert cases like trying to extend the dates by not the lessor, etc

File: contracts/Rentals.sol (hunk 1 above)
Comment (fzavalia) [reply contracts/Rentals.sol:412]: It's kind of complicated to read

File: test/Rentals.spec.ts (hunk 4 above)
Comment (fzavalia) [reply test/Rentals.spec.ts:2264]: Added various tests for extend revertions

File: test/Rentals.spec.ts [hunk 5]
```diff
@@ -1338,6 +1431,81 @@ describe('Rentals', () => {
...
+
+      await evmMine()
+
+      const latestBlockTimestamp = await getLatestBlockTimestamp()
+
+      rental = await rentals.rentals(listingParams.contractAddress, listingParams.tokenId)
+
+      expect(rental.lessor).to.equal(lessor.address)
+      expect(rental.tenant).to.equal(tenant.address)
+      expect(rental.endDate).to.equal(latestBlockTimestamp + daysToSeconds(rentalDays1 + rentalDays2))
```
Comment (nachomazzara) [inline test/Rentals.spec.ts:1554]: Can we check at this instance the lessor and tenant nonce?

File: test/Rentals.spec.ts [hunk 6]
```diff
@@ -1338,6 +1431,81 @@ describe('Rentals', () => {
...
+      expect(rental.tenant).to.equal(zeroAddress)
+      expect(rental.endDate).to.equal(0)
+
+      await network.provider.send('evm_setAutomine', [false])
+
+      offerParams = { ...offerParams, rentalDays: rentalDays1 }
+
+      await rentals.connect(lessor).acceptOffer({ ...offerParams, signature: await getOfferSignature(tenant, rentals, offerParams) })
+
+      offerParams = { ...offerParams, rentalDays: rentalDays2, nonces: [0, 0, 1] }
```
Comment (nachomazzara) [inline test/Rentals.spec.ts:1540]: Can we check at this instance the lessor and tenant nonce?

File: test/Rentals.spec.ts [hunk 7]
```diff
@@ -1338,6 +1431,81 @@ describe('Rentals', () => {
       expect(rental.endDate).to.equal(endDate)
     })
 
+    it('should allow calling accept offer twice if the second is an extension in the same block', async () => {
+      const rentalDays1 = 10
+      const rentalDays2 = 100
+
+      let rental = await rentals.rentals(listingParams.contractAddress, listingParams.tokenId)
```
Comment (nachomazzara) [inline test/Rentals.spec.ts:1528]: Can we check at this instance the lessor and tenant nonce?

File: test/Rentals.spec.ts [hunk 8]
```diff
@@ -1338,6 +1431,81 @@ describe('Rentals', () => {
...
+      const latestBlockTimestamp = await getLatestBlockTimestamp()
+
+      rental = await rentals.rentals(listingParams.contractAddress, listingParams.tokenId)
+
+      expect(rental.lessor).to.equal(lessor.address)
+      expect(rental.tenant).to.equal(tenant.address)
+      expect(rental.endDate).to.equal(latestBlockTimestamp + daysToSeconds(rentalDays1 + rentalDays2))
+    })
+
+    it('should allow calling accept offer then an accept listing if the second is an extension in the same block', async () => {
```
Comment (nachomazzara) [inline test/Rentals.spec.ts:1557]: ❤️

File: test/Rentals.spec.ts [hunk 9]
```diff
@@ -1338,6 +1431,81 @@ describe('Rentals', () => {
...
+      await evmMine()
+
+      const latestBlockTimestamp = await getLatestBlockTimestamp()
+
+      rental = await rentals.rentals(listingParams.contractAddress, listingParams.tokenId)
+
+      expect(rental.lessor).to.equal(lessor.address)
+      expect(rental.tenant).to.equal(tenant.address)
+      expect(rental.endDate).to.equal(latestBlockTimestamp + daysToSeconds(rentalDays1 + rentalDays2))
+    })
```
Comment (nachomazzara) [inline test/Rentals.spec.ts:1597]: I would check again the nonces and rentals properties before and after each tx

File: test/Rentals.spec.ts [hunk 10]
```diff
@@ -603,6 +603,99 @@ describe('Rentals', () => {
...
+
+      await evmMine()
+
+      const latestBlockTimestamp = await getLatestBlockTimestamp()
+
+      rental = await rentals.rentals(listingParams.contractAddress, listingParams.tokenId)
+
+      expect(rental.lessor).to.equal(lessor.address)
+      expect(rental.tenant).to.equal(tenant.address)
+      expect(rental.endDate).to.equal(latestBlockTimestamp + daysToSeconds(rentalDays1 + rentalDays2))
```
Comment (nachomazzara) [inline test/Rentals.spec.ts:656]: I would check again the nonces and rentals properties before and after each tx

File: test/Rentals.spec.ts [hunk 11]
```diff
@@ -603,6 +603,99 @@ describe('Rentals', () => {
...
```
Comment (nachomazzara) [inline test/Rentals.spec.ts:698]: I would check again the nonces and rentals properties before and after each tx

File: test/Rentals.spec.ts [hunk 12]
```diff
@@ -1966,6 +2134,87 @@ describe('Rentals', () => {
...
```
Comment (nachomazzara) [inline test/Rentals.spec.ts:2344]: Can we add also tests when the block is not minted but the sender is trying to reuse the signature?

File: test/Rentals.spec.ts (hunk 9 above)
Comment (fzavalia) [reply test/Rentals.spec.ts:1725]: Done

File: test/Rentals.spec.ts (hunk 7 above)
Comment (fzavalia) [reply test/Rentals.spec.ts:1592]: Checked on all similar extension in same block tests

File: test/Rentals.spec.ts (hunk 12 above)
Comment (fzavalia) [reply test/Rentals.spec.ts:2521]: https://github.com/decentraland/rentals-contract/pull/23/files#diff-3369bd7d295f9e1649f509622263f562231b444e193ccbdd316915f134efd28fR1404 That is verified here

Comment (fzavalia) [reply test/Rentals.spec.ts:2521]: There are various tests like this for accept offer, accept listing and accepting offer by sending estate and land in different combinations

**Pull request merged**